### PR TITLE
One Lolly CTL for World of Padman

### PR DESCRIPTION
**Pull request by kai-li-wop**
This is currently a draft and work in progress to try to implement **One Lolly Capture The Lolly (1LCTL)** as a new game type to World of Padman. Most of the code is from Quake3 Team Arena, the assets are original WoP assets but adapted to work with 1LCTL or newly created. Some of it is also just a placeholder. Since Team Arena works a bit different and WoP does not use or contain the Team Arena code, some things need to be done differently or recreated. So help and support is more than welcome. This could be released as a modification for the upcoming World of Padman release or as part of it if accepted by the team.

1LCTL uses the default CTL HUD at the lower right but in a modified way. The upper lolly slot show a blue or red lolly icon (@) depending on the neutral lolly taken status to indicate which team has currently taken the neutral lolly. The lower lolly slot shows different white lolly icons depending on the neutral lolly status; at base (@), taken (x) and dropped (?). I also experimented with only one lolly slot for 1LCTL, but I disliked the outcome.

ToDo
- [x] Transfer and adjust 1FCTF code from Q3TA
- [x] Adjust the CTL HUD to work with 1LCTL
- [x] Provide green lolly icons and shader for the neutral lolly
- [x] Adjust Team Overlay and Scoreboard for 1LCTL 
- [x] Provide a lolly model with white textures
- [ ] Provide necessary sound samples --> #19
- [x] Provide a test map (test with Harm's Westernset)
- [x]  Provide a map source file in XTRAS folder
- [x] Adjust bot AI for 1LCTL --> #261
- [x] Adjust WoP's CTL maps for 1LCTL (except ENTE's maps)

### CODE REVIEW

File: code/game/g_team.c [hunk 1]
```diff
@@ -643,32 +664,35 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
 	gclient_t *cl = other->client;
 	int enemy_flag;
 
-	if (cl->sess.sessionTeam == TEAM_RED) {
-		enemy_flag = PW_BLUEFLAG;
+	if(g_gametype.integer == GT_1FCTF) {
```
Comment (mgerhardy) [inline code/game/g_team.c]: missing whitespace

File: code/game/g_team.c [hunk 2]
```diff
@@ -643,32 +664,35 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
...
+		}
 	}
-
 	// the flag is at home base.  if the player has the enemy
 	// flag, he's just won!
 	if (!cl->ps.powerups[enemy_flag])
 		return 0; // We don't have the flag
-	PrintMsg(NULL, "%s" S_COLOR_WHITE " captured the %s' lolly!\n", cl->pers.netname, TeamName(OtherTeam(team)));
-
+	if( g_gametype.integer == GT_1FCTF ) {
```
Comment (mgerhardy) [inline code/game/g_team.c]: formatting is off

File: code/game/g_team.c [hunk 3]
```diff
@@ -728,14 +752,28 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
...
+	if(g_gametype.integer == GT_1FCTF) {
+		PrintMsg (NULL, "%s" S_COLOR_WHITE " got the lolly!\n", other->client->pers.netname );
 
-	if (team == TEAM_RED)
-		cl->ps.powerups[PW_REDFLAG] = INT_MAX; // flags never expire
-	else
-		cl->ps.powerups[PW_BLUEFLAG] = INT_MAX; // flags never expire
+		cl->ps.powerups[PW_NEUTRALFLAG] = INT_MAX; // flags never expire
+
+		if( team == TEAM_RED ) {
```
Comment (mgerhardy) [inline code/game/g_team.c]: formatting

File: code/game/g_items.c [hunk 4]
```diff
@@ -728,11 +728,29 @@ void G_CheckTeamItems(void) {
...
 		}
 		item = BG_FindItem("blue Lolly");
 		if (!item || !itemRegistered[item - bg_itemlist]) {
-			G_Printf(S_COLOR_YELLOW "WARNING: No team_CTL_bluelolly in map");
+			G_Printf(S_COLOR_YELLOW "WARNING: No team_CTL_bluelolly in map\n");
+		}
+	}
+
+	if( g_gametype.integer == GT_1FCTF ) {
+		gitem_t	*item;
```
Comment (mgerhardy) [inline code/game/g_items.c]: this must be a `const` pointer

File: code/game/g_items.c [hunk 5]
```diff
@@ -728,11 +728,29 @@ void G_CheckTeamItems(void) {
...
+
+		// check for all three flags
+		item = BG_FindItem( "Red Lolly" );
```
Comment (mgerhardy) [inline code/game/g_items.c]: formatting is off here

File: code/botlib/be_ai_goal.c [hunk 6]
```diff
@@ -89,7 +89,8 @@ typedef struct campspot_s {
 typedef enum {
 	BOTLIB_GT_SINGLE_PLAYER,
 	BOTLIB_GT_FFA,
-	BOTLIB_GT_TEAM
+	BOTLIB_GT_TEAM,
+	BOTLIB_GT_1FCTF,
```
Comment (mgerhardy) [inline code/botlib/be_ai_goal.c]: please remove the trailing `,` after `BOTLIB_GT_1FCTF`

File: code/cgame/cg_draw.c [hunk 7]
```diff
@@ -1323,7 +1323,7 @@ static void CG_DrawDisconnect(void) {
 		x = (640 - 48 - 94);
 	} else if (cgs.gametype == GT_BALLOON) {
 		x = (640 - 48 - 64);
-	} else if (cgs.gametype == GT_CTF) {
+	} else if ((cgs.gametype == GT_CTF) || (cgs.gametype == GT_1FCTF)) {
```
Comment (mgerhardy) [inline code/cgame/cg_draw.c]: `if (cgs.gametype == GT_CTF || cgs.gametype == GT_1FCTF)` this is enough - no extra `()` needed

File: code/cgame/cg_main.c [hunk 8]
```diff
@@ -1047,7 +1062,16 @@ static void CG_RegisterGraphics(void) {
 			cgs.media.bambamHealthIcon = trap_R_RegisterShaderNoMip("models/weapons2/bambam/bamhealth02");
 		}
 
-		if ((cgs.gametype == GT_CTF) || (cgs.gametype == GT_BALLOON) || cg_buildScript.integer) {
+		if ( cgs.gametype == GT_1FCTF || cg_buildScript.integer ) {
```
Comment (mgerhardy) [inline code/cgame/cg_main.c]: formatting is wrong here.

File: code/cgame/cg_predict.c [hunk 9]
```diff
@@ -291,6 +291,12 @@ static void CG_TouchItem(centity_t *cent) {
 
 	// Special case for flags.
 	// We don't predict touching our own flag
+	if(cgs.gametype == GT_1FCTF) {
```
Comment (mgerhardy) [inline code/cgame/cg_predict.c]: formatting is wrong here

File: code/game/bg_misc.c [hunk 10]
```diff
@@ -776,6 +790,22 @@ qboolean BG_CanItemBeGrabbed(int gametype, const entityState_t *ent, const playe
 		return qtrue; // powerups are always picked up
 
 	case IT_TEAM: // team items, such as flags
+		if (gametype == GT_1FCTF) {
+			// neutral flag can always be picked up
+			if(item->giTag == PW_NEUTRALFLAG) {
```
Comment (mgerhardy) [inline code/game/bg_misc.c]: formatting is wrong here `if(` should be `if (`

File: code/game/bg_misc.c [hunk 11]
```diff
@@ -776,6 +790,22 @@ qboolean BG_CanItemBeGrabbed(int gametype, const entityState_t *ent, const playe
...
+				return qtrue;
+			}
+			if (ps->persistant[PERS_TEAM] == TEAM_RED) {
+				if (item->giTag == PW_BLUEFLAG  && ps->powerups[PW_NEUTRALFLAG]) {
```
Comment (mgerhardy) [inline code/game/bg_misc.c]: formatting is wrong here - `  && ` should be ` && `

File: code/game/g_items.c [hunk 12]
```diff
@@ -728,11 +728,29 @@ void G_CheckTeamItems(void) {
...
+			G_Printf(S_COLOR_YELLOW "WARNING: No team_CTL_redlolly in map\n");
 		}
 		item = BG_FindItem("blue Lolly");
 		if (!item || !itemRegistered[item - bg_itemlist]) {
-			G_Printf(S_COLOR_YELLOW "WARNING: No team_CTL_bluelolly in map");
+			G_Printf(S_COLOR_YELLOW "WARNING: No team_CTL_bluelolly in map\n");
+		}
+	}
+
+	if(g_gametype.integer == GT_1FCTF) {
```
Comment (mgerhardy) [inline code/game/g_items.c]: missing whitespace between if and (

File: code/game/g_team.c [hunk 13]
```diff
@@ -520,6 +533,10 @@ static void Team_ResetFlags(void) {
 		Team_ResetFlag(TEAM_RED);
 		Team_ResetFlag(TEAM_BLUE);
 	}
+
```
Comment (mgerhardy) [inline code/game/g_team.c]: else if should be in the same line as the closing brace }

File: code/game/g_team.c [hunk 14]
```diff
@@ -728,14 +752,28 @@ static int Team_TouchOurFlag(gentity_t *ent, gentity_t *other, int team) {
...
 
-	if (team == TEAM_RED)
-		cl->ps.powerups[PW_REDFLAG] = INT_MAX; // flags never expire
-	else
-		cl->ps.powerups[PW_BLUEFLAG] = INT_MAX; // flags never expire
+		cl->ps.powerups[PW_NEUTRALFLAG] = INT_MAX; // flags never expire
+
+		if (team == TEAM_RED) {
+			Team_SetFlagStatus(TEAM_FREE, FLAG_TAKEN_RED);
+		}
```
Comment (mgerhardy) [inline code/game/g_team.c]: else should be in the same line as the closing }

**Review by mgerhardy — COMMENTED** [review]
nice job so far - mostly about cosmetic formatting stuff. this is only "important" because the next clang-format run would produce a lot of diffs otherwise.

**Comment by kai-li-wop** [comment]
Most of the stuff is simply copy paste from Team Arena and the formatting issues of TA code are not easy to spot. Is there a VS Code plugin for this? Any suggestions?

**Pull request merged**